x/tally/keeper/endblock.go

```diff
@@ -227,14 +254,48 @@ func (k Keeper) FilterAndTally(ctx sdk.Context, req types.Request, params types.
 		}
 	}
 
-	// Phase III: Calculate Payouts
-	// TODO: Calculate gas used & payouts
-	result.ExecGasUsed = calculateExecGasUsed(reveals)
-	distMsgs := types.DistributionMessages{
-		Messages:   []types.DistributionMessage{},
-		RefundType: types.DistributionTypeNoConsensus,
+	// Phase 3: Calculate Payouts
+	// Calculate data proxy payouts if basic consensus was reached.
+	var proxyDistMsgs []types.Distribution
+	var proxyGasUsedPerExec uint64
+	if filterErr == nil || !errors.Is(filterErr, types.ErrNoBasicConsensus) {
+		proxyDistMsgs, proxyGasUsedPerExec = k.CalculateDataProxyPayouts(ctx, result.ProxyPubKeys, gasPrice)
+	}
+
+	// Calculate executor payouts.
+	var execDistMsgs []types.Distribution
+	execBurnAmt := math.NewInt(0)
+	reducedPayout := false
+	switch {
+	case errors.Is(filterErr, types.ErrNoBasicConsensus):
+		execDistMsgs = k.CalculateCommitterPayouts(ctx, req, params.GasCostCommit, gasPrice)
+	case errors.Is(filterErr, types.ErrInvalidFilterInput) || errors.Is(filterErr, types.ErrNoConsensus) || tallyErr != nil:
+		reducedPayout = true
+		fallthrough
+	default: // filterErr == ErrConsensusInError || filterErr == nil
+		gasReports := make([]uint64, len(reveals))
+		for i, reveal := range reveals {
+			if proxyGasUsedPerExec > reveal.GasUsed {
+				gasReports[i] = 0
+			} else {
+				gasReports[i] = reveal.GasUsed - proxyGasUsedPerExec
+			}
+		}
+
+		var execGasUsed uint64
+		if req.ReplicationFactor == 1 || areGasReportsUniform(gasReports) {
```

Comment:
Can we not add the RF=1 case to `areGasReportsUniform`? Since we check for 0 there as well.

Reply:
Changed the check in `areGasReportsUniform` to 
```
if len(reports) <= 1 {
		return true
}
```